Share reason why target is no longer visible

lread opened the pull request:
A var that:
- has become private is logged as an error.
- no longer includes metadata specified by `:exclude-meta`
  is logged as a warning.

To support this change:
- factored out println logging to log-diff fn
  - re-worked arity error msg for consistency
- added flatmap dependency for its ordered map
  - result ordering should now actually be consistent

Closes #13

lread committed:
Share reason why target is no longer visible

A var that:
- has become private is logged as an error.
- no longer includes metadata specified by `:exclude-meta`
  is logged as a warning.

To support this change:
- factored out println logging to log-diff fn
  - re-worked arity error msg for consistency
- added flatmap dependency for its ordered map
  - result ordering should now actually be consistent

Closes #13

lread commented:
Lemme know whatcha think!
Adding flatland ordered-map work for you?

borkdude commented:
Going to take a deeper look tomorrow, but at first glance it looks pretty sweet! Could you clarify why you need a sorted map? Preferably I'd do with the least amount of deps, but sometimes convenience wins and flatland/ordered-map is not that large.

lread commented:
Hiya!

We want to spit out our findings in the same order every time. 
I figured sorting by ns then row was a good order, and thought I had achieved this with a previous change but missed the mark.

Switching to the sorted map is a small tweak to the code to ensure consistent report sort order using the same basic data structures.

We could achieve the same without a sorted map but would have to restructure or come up with another way to achieve the same sort order. Certainly not rocket science, but also not necessary if you are cool with bringing this new dep.

As usual, pros and cons!

lread commented:
And I wonder if that flatmap dep will be permanent? Maybe not after tackling #14 which might benefit from a rethink of the comparison algorithm. But for now, it is useful.

borkdude commented:
> And I wonder if that flatmap dep will be permanent? Maybe not after tackling #14 which might benefit from a rethink of the comparison algorithm. But for now, it is useful.

I'm not convinced we need this dependency enough. It's limited to the `index-by` function which is only used to create a group-by map which we `doseq` over later. Since we doseq over it we don't actually need the map. I only did it like this because it was convenient, not because it was the right data structure. Instead the `group` function can just create a seq of tuples for example. An alternative is to sort the data after all the findings have been emitted to an atom for example. This is what is done in clj-kondo.

lread commented:
Good to know @borkdude, thanks.
We can certainly do away with the extra dep then, I shall make it so!

lread committed:
Respond to feedback: remove flatland dependency

While moving to drive the compare from a map to sequence,
I stumbled on the fact clj-kondo var-definitions
contain separate entries for declares.

Any declares entries are now excluded from the diff.
